### DISCUSSION TITLE
mmix does not work properly in INAV 4.0

### DISCUSSION THREAD

**Discussion by ABLomas** [iNavFlight/inav#7819] [Q&A]
With 4.0 release something happen with GUI control and custom mixer. Downloaded fresh copy of 4.0 configurator, installed copy to kakute F7.
OK, just trying to swap motors: https://t.me/INAVFlight/220834 - it reverts after reboot without any effect.
Going to do the same with mixers (swapping control axis in GUI, in other words, putting "-1" for 1 to swap motors) has no effect.
One guy in INAV TG said that doing the same via CLI does not work, too (https://t.me/INAVFlight/221196), while "Miel'O'txin" said, that motor wizard is working for him.
Earlier i could do just do something like:
`mixer custom
mmix reset
mmix ....`
in CLI alone, but now "mixer" command is unavailable and documentation (https://github.com/iNavFlight/inav/blob/master/docs/Mixer.md) does not explain how to swap different mixers (for example, from QUADX to QUAD+ or get current mixer modified).

I just re-tried to do custom mmix in CLI with swapped axis, after 'save' and reboot 'mmix' command output is empty. Loaded mixer via GUI - mmix lists custom values, but motor wizard again does not work. 

So, in short - changing motor order does not work on some platforms and there is no (clear/documented) way to work around this issue in CLI.

**Comment by stronnag** [comment]
see  https://github.com/iNavFlight/inav/wiki/Custom-mixes-for-exotic-setups

custom mixer works fine for my vtail, last time I tried it.

```
# mixer
mmix 0  1.000 -0.580  0.580  1.000
mmix 1  1.000 -0.460 -0.390 -0.500
mmix 2  1.000  0.580  0.580 -1.000
mmix 3  1.000  0.460 -0.390  0.500
```

**Comment by stronnag** [comment]
See also 
https://github.com/iNavFlight/inav/wiki/Legacy-Mixers

which may also explain some of the misconceptions.

**Comment by ABLomas** [comment]
Agree, that is may work, but let's try - here is default QuadX mixer:
```
# mmix
mmix 0  1.000 -1.000  1.000 -1.000
mmix 1  1.000 -1.000 -1.000  1.000
mmix 2  1.000  1.000  1.000  1.000
mmix 3  1.000  1.000 -1.000 -1.000
```

Tested by working in hands - due to my copter being inverted configuration - motors need to be swapped - so it stabilizes in wrong order.
OK, let's enter those lines:
```
mmix 0  1.000 1.000  -1.000 -1.000
mmix 1  1.000 1.000 1.000  1.000
mmix 2  1.000 -1.000 -1.000  1.000
mmix 3  1.000 -1.000 1.000 -1.000
```

So final mixer becomes:
```
# mmix
mmix 0  1.000  1.000 -1.000 -1.000
mmix 1  1.000  1.000  1.000  1.000
mmix 2  1.000 -1.000 -1.000  1.000
mmix 3  1.000 -1.000  1.000 -1.000
```
(maybe wrong for actual configuration, but now should invert axis stabilisation for sure)
...save
...and here i have the same stabilisation (if rotate copter in hands - the same motors are speeding up as in QuadX config).

I know, this is working for some, but here is one setup which is not working OK on 4.0 for me (starting from the same motor wizard, i linked video above).

**Comment by ekap12** [comment]
my video maybe same problem mixers does not work https://youtu.be/Or3Ru7hi0IE
Matek 722 std
hobbywing XRotor Micro 60A
In betaflight remaping work fine
in "motor reverse" option motors does not work

**Comment by DzikuVx** [comment]
It is working for all cases. There is a misconception: custom mixer does not change the motor order. Custom mixer changes the mixer. Motor connected to S1 will still be connected to S1, S2 will be S2 and so on.

**Reply by ABLomas** [reply]
You stated "it is working in all cases" - well, did you saw video (first link in first post)?
It is working here as intended?

Also, i understand that custom mmix does not change motor order. If motor was speeding when tilting on that corner, then with inverted mixer it should slow down. Instead, there is no difference on motor output - it is speeding up like nothing happen.

**Reply by Odin263** [reply]
> You stated "it is working in all cases" - well, did you saw video (first link in first post)? It is working here as intended?
> 
Yes, it also works as it should in the videos.
see here:
before the change
![image](https://user-images.githubusercontent.com/90968252/153498406-40e172c1-2d03-42c5-b843-c5785480424c.png)
Motor 1 =  1,-1, 1,-1
Motor 2 =  1,-1,-1, 1
Motor 3 =  1, 1, 1, 1
Motor 4 =  1, 1,-1,-1

after the change
![image](https://user-images.githubusercontent.com/90968252/153499056-69938d78-6879-4ae9-9b41-7160fd024cdb.png)
Motor 1 (3) =  1, 1, 1, 1
Motor 2 (4) =  1, 1,-1,-1
Motor 3 (1) =  1,-1, 1,-1
Motor 4 (2) =  1,-1,-1, 1

To understand why there is no difference in balancing direction:
Before the change, motor 1 (rear-right) for roll is defined as -1. That means when accelerating the engine, the quad rotates counterclockwise when rolling, which would also be correct if the position of the motor were correct.
However, since the engine is on the other side (rear-left), the definition changes to 1 for roll (clockwise), which is still correct.

As you can see, it remains correct if the position of the motors was initially assumed to be wrong. Likewise, it remains wrong even if it was previously defined incorrectly.
Why should the behavior change? Only the position of the motors should be mentioned.

The easiest way is to set the mixer correctly according to the picture
![image](https://user-images.githubusercontent.com/90968252/153503587-c47bdf39-226b-473b-a1d0-f745e45fa649.png)
and then adjust the position of the motors if necessary.
